examples/twittermap/web/public/javascripts/timeseriescache/services.js

```diff
@@ -0,0 +1,128 @@
+/*
+ * This module caches time-series histogram data. The common services module communicates with
+ * middle ware and the time-series cache.
+ */
+'use strict';
+angular.module('cloudberry.timeseriescache', [])
+    .service('TimeSeriesCache', function () {
+
+        // The key-value stores time series results of a query.
+        var timeseriesStore = new HashMap();
+        // To check if keyword in query is changed (new query?)
+        var currentKeywords = [""];
+        // To check if geo level in query is changed.
+        var currentGeoLevel = "";
+        // To check if time range in query changed (new query?)
+        var endDate = new Date();
+        // Deducts -1 to create a default value that can't exist
+        endDate.setDate(endDate.getDate() - 1);
+        // Cached time interval of complete byTimeRequest result.
+        var cachedTimeRange = {
+            start: new Date(),
+            end: endDate
+        };
+        const INVALID_VALUE = 0;
+
+        /**
+         * Checks keyword, time range and the cache store, and returns the geoIds that
+         * are not already cached.
+         */
+        this.getGeoIdsNotInCache = function (keywords, timeInterval, geoIds, geoLevel) {
+            // The length of geoIdsNotInCache is 0 in case of complete cache hit,
+            // same length as the geoIds parameter in case of complete cache miss,
+            // otherwise in range (0, geoIds.length)
+            var geoIdsNotInCache = [];
+            // New query case
+            if (keywords.toString() != currentKeywords.toString() ||
+                timeInterval.start < cachedTimeRange.start ||
+                timeInterval.end > cachedTimeRange.end ||
+                geoLevel != currentGeoLevel) {
+                timeseriesStore.clear();
+                currentKeywords = keywords.slice();
+                currentGeoLevel = geoLevel;
+                cachedTimeRange.start = timeInterval.start;
+                cachedTimeRange.end = timeInterval.end;
+
+                return geoIds;
+            }
+            for (var i = 0; i < geoIds.length; i++) {
+                if (!timeseriesStore.has(geoIds[i])) {
+                    geoIdsNotInCache.push(geoIds[i]);
+                }
+            }
+
+            return geoIdsNotInCache;
+        };
+
+        /**
+         * Retrieves time-series data from the cache; ignores empty objects
+         */
+        this.getTimeSeriesValues = function (geoIds, geoLevel, timeInterval) {
```

Comment:
It seems like it returns all day and count and ignores geoIDs. Why are geoIDs not required? Also, this will return the same day multiple times per GeoID. Is that so?

Reply:
[line64] var values = timeseriesStore.get(geoIds[i]);
This returns one {day:count} per GeoID, but may return records of the same date for different GeoIDs. (I decided not to merge the count in this case because the histogram library handles it pretty well)

Reply:
OK.